Add expand_labels function to dilate segments while avoiding overlaps

VolkerH opened the pull request:
## Description

This implements growing of connected component regions in a label array up to a specified number of pixels/voxels, but not starting to merge/overlap separate regions.

This functionality is similar to the `Distance N` method in Cellprofiler's IdentifySecondaryObjects module, in fact I adapted the code from [there](
https://github.com/CellProfiler/CellProfiler/blob/022c55f3c3644eccbd285ae1bf8ed242d3416b2b/cellprofiler/modules/identifysecondaryobjects.py#L569) but modified it to support arbitrary dimensions (that should tick a box with @jni).  Also see the discussion on [image.sc](https://forum.image.sc/t/equivalent-to-cellprofilers-identifysecondaryobjects-distance-n-in-fiji/39146).
I am not aware of an original paper I can cite for this, but there probably is one.

I created a demo notebook that shows what this is doing. You can see the demo notebook [here](https://github.com/VolkerH/PythonSnippets/blob/master/ExpandLabels/expand_labels_demo.ipynb).

This is still a WIP.
Tests, examples, and formatting still to do.

## Checklist

- (started) [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- (TODO) Gallery example in `./doc/examples` (new features only)
- (TODO) Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- (TODO) Unit tests
- (TODO) Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)


## For reviewers

<!-- Don't remove the checklist below. -->
- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

pep8speaks commented:
Hello @VolkerH! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`skimage/measure/_label.py`](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py):

> [Line 98:80](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L98): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 100:71](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L100): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 102:71](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L102): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 103:74](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L103): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 110:67](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L110): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 112:1](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L112): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 113:80](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L113): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 113:85](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L113): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 114:80](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L114): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 114:85](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L114): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 115:80](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L115): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 116:54](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L116): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 118:80](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L118): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 120:79](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L120): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 139:80](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L139): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 79 characters)
> [Line 140:80](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L140): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 79 characters)
> [Line 146:1](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L146): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 147:76](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L147): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 147:78](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L147): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 147:80](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L147): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 150:80](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L150): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 157:1](https://github.com/scikit-image/scikit-image/blob/5eebd72ad5ac4c38b9313644ce42677470397c6b/skimage/measure/_label.py#L157): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

VolkerH commented:
A question I have is:

Should this go into `skimage.measure` (I chose that because that's where `label` lives) or preferably somewhere else ?

VolkerH commented:
Tests, documentation and gallery example have been added with contributions from @jni.
This is now ready for review.

jni commented:
@scikit-image/core This is ready for review! Test failures are:

- sphinx link error in the See Also section (working to fix it)
- matplotlib 3.3.0rc1 breaking things

For reference, here is the output of the gallery example:

![expand_labels](https://user-images.githubusercontent.com/492549/85371339-abc4a580-b4f5-11ea-9168-8c4a838e9757.png)

VolkerH commented:
unrelated to this pull request, but looking at the example screenshot. What is the issue with `label2rgb`  that makes the overlay of the background greyish for the dilated labels, something to do with automated contrast scaling?

jni commented:
> What is the issue with label2rgb that makes the overlay of the background greyish for the dilated labels

I expect we've done something wrong with the background label, which is supposed to be transparent...

jni commented on doc/examples/segmentation/plot_expand_labels.py:
```diff
@@ -0,0 +1,54 @@
+axes[0].imshow(color1)
+axes[0].set_title('Sobel+Watershed')
+
+color2 = label2rgb(expanded, image=coins, image_alpha=0.5, bg_label=0)
```
Ah, here's the issue, image_alpha should be removed. Will do that now.

jni replied:
Actually: I'll do that after I see if the build works.

VolkerH replied:
Thanks, so it was actually related to our example.

jni commented:
Fixed! This is now *definitely* ready for review, @scikit-image/core! =) Again, the remaining failure is unrelated to this PR and due to the matplotlib pre-release.

rfezzani commented on skimage/segmentation/tests/test_expand_labels.py:
```diff
@@ -0,0 +1,189 @@
+    (EDGECASE2D, EDGECASE2D_EXPANDED_4, 4),
+    (SAMPLE3D, SAMPLE3D_EXPANDED_2, 2)
+    ]
+
```
```suggestion
```

rfezzani commented on skimage/segmentation/tests/test_expand_labels.py:
```diff
@@ -0,0 +1,189 @@
+    assert_array_equal(expanded, expected_output)
+
+
+@testing.parametrize(
+    'ndim, distance', list(product([2, 3], range(6))),
+)
```
I think that you can chain the decorators and avoid the use/import of `product`:
```suggestion
@testing.parametrize('ndim', [2, 3])
@testing.parametrize('distance', range(6))
```

rfezzani commented on skimage/segmentation/tests/test_expand_labels.py:
```diff
@@ -0,0 +1,189 @@
+    expanded = expand_labels(SAMPLE_EDGECASE_BEHAVIOUR, 1)
+    expanded_transpose = expand_labels(SAMPLE_EDGECASE_BEHAVIOUR.T, 1)
+    assert not np.all(expanded == expanded_transpose.T)
+
+     
```
Unnecessary extra blank lines:
```suggestion
```

VolkerH replied:
removed, thanks

VolkerH replied:
for some reason github did not let me commit your suggestion here, so I had to do it in my text editor and push the change. I don't quite get why ...

rfezzani commented on skimage/segmentation/tests/test_expand_labels.py:
```diff
@@ -0,0 +1,189 @@
+    beyond_expanded_distances = distance_map[~expanded.astype(bool)]
+    if beyond_expanded_distances.size > 0:
+        assert np.all(beyond_expanded_distances > distance)
+
```
Extra blank line required (pep8 :slightly_smiling_face: )
```suggestion


```

rfezzani commented on skimage/segmentation/tests/test_expand_labels.py:
```diff
@@ -0,0 +1,189 @@
+    - Distance beyond the expanded labels should always exceed the input
+      distance.
+    """
+    array = data.binary_blobs(length=64, blob_size_fraction=0.05, n_dim=ndim)
```
`img` or `image` is may be a better name

VolkerH replied:
agreed, just pushed that change.

rfezzani commented on skimage/segmentation/tests/test_expand_labels.py:
```diff
@@ -0,0 +1,189 @@
+from skimage._shared.testing import assert_array_equal
+
+SAMPLE1D = np.array([0, 0, 4, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0])
+SAMPLE1D_EXPANDED_3 = array([4, 4, 4, 4, 4, 4, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0])
```
Please be consistent with previous line ;)
```suggestion
SAMPLE1D_EXPANDED_3 = np.array([4, 4, 4, 4, 4, 4, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0])
```

VolkerH replied:
thanks for spotting this copy&paste error

rfezzani commented on skimage/segmentation/tests/test_expand_labels.py:
```diff
@@ -0,0 +1,189 @@
+     [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0]]
+)
+
+EDGECASE2D_EXPANDED_4 = array(
```
```suggestion
EDGECASE2D_EXPANDED_4 = np.array(
```

VolkerH commented on skimage/segmentation/_expand_labels.py:
```diff
@@ -0,0 +1,96 @@
+    ----------
+    label_image : ndarray of dtype int
+        label image
+    distance : float
```
doesn't have to be as the distance transform returns floating point, there is even a test for expanding by 1.5

VolkerH commented on skimage/segmentation/_expand_labels.py:
```diff
@@ -0,0 +1,96 @@
+    ----------
+    .. [1] https://cellprofiler.org
+    .. [2] https://github.com/CellProfiler/CellProfiler/blob/082930ea95add7b72243a4fa3d39ae5145995e9c/cellprofiler/modules/identifysecondaryobjects.py#L559
+    .. [3] https://forum.image.sc/t/equivalent-to-cellprofilers-identifysecondaryobjects-distance-n-in-fiji/39146/16
```
I originally linked to it as the linked forum thread involved a discussion of the edge cases and how they can be a biasing the results. 
Now that the edge cases are discussed in the docstring (although not the issue of bias) I agree this can be removed.

VolkerH commented on skimage/measure/_label.py:
```diff
@@ -1,5 +1,16 @@
+All rights reserved.
+
+Original author/s: Cellprofiler team 
+"""
```
moved

VolkerH commented on skimage/segmentation/_expand_labels.py:
```diff
@@ -0,0 +1,96 @@
+    There is an important edge case when a pixel has the same distance to
+    multiple regions, as it is not defined which region expands into that
+    space, see the discussion in [1]. Here, the exact bahaviour depends on
+    the upstream implementation.
```
Thanks, I added a Notes section to the docstring and moved the paragraph there.

VolkerH commented:
Thanks for spotting all of these @rfezzani, I think I addressed all of them (starting to lose the overview).

rfezzani commented on skimage/segmentation/tests/test_expand_labels.py:
```diff
@@ -0,0 +1,189 @@
+from skimage import data
+
+import numpy as np
+from numpy import array
```
This import is no more needed ^^
```suggestion
```

rfezzani commented on skimage/segmentation/tests/test_expand_labels.py:
```diff
@@ -0,0 +1,189 @@
+        [0, 0, 0, 0],
+        [0, 0, 5, 0]]])
+
+SAMPLE3D_EXPANDED_2 =np.array(
```
```suggestion
SAMPLE3D_EXPANDED_2 = np.array(
```

stefanv commented on doc/examples/segmentation/plot_expand_labels.py:
```diff
@@ -0,0 +1,54 @@
+Expand segmentation labels without overlap
+==========================================
+
+When you have a set of regions
```
Starts with half a sentence.

VolkerH replied:
Somehow that docstring still had the pasted text from a different example. Should be better now.

stefanv commented on skimage/segmentation/_expand_labels.py:
```diff
@@ -0,0 +1,107 @@
+""" 
+expand_labels is  inspired by code in CellProfiler, 
```
All licensing information should please go into LICENSE.txt.

VolkerH replied:
Including this in the file was suggested by @jni who pointed out that this kind of note is also present in e.g. `_canny.py`, `_edges.py`, `_watershed.py` and a number of other algorithms that were also originally implemented for CellProfiler.

Without digging too deep License.txt seems to contain copyright notices for whole files that have benn vendored?

I can move the note elsewhere (License.txt if that is what the core team agrees on), but I'd rather err on the side of attributing the original authors one time more than necessary.

VolkerH replied:
I removed the notice from `_expand_labels.py` for now but reworded the attribution more clearly in the Notes section. We probably still need to retain the original string somewhere.
Maybe all the different Cellprofiler contributions could be acknowledged in a bundled form in License.txt for consistency. I'm not sure whether that is in agreement with the actual terms of the BSD-3 license. If so, this would be a refactor that should probably be in a separate PR anyway (and probably with input from the Cellprofiler team to make sure they are happy).

stefanv replied:
I wasn't sure whether "inspired" meant derived from.  We have to decide if this is a copyright statement or not, because that influences under which license we release the file (CallProfiler's, or ours).

VolkerH replied:
I had changed that several times before. The core approach is based on the lines following this: https://github.com/CellProfiler/CellProfiler/blob/082930ea95add7b72243a4fa3d39ae5145995e9c/cellprofiler/modules/identifysecondaryobjects.py#L559 however, the final result has quite a few differences most importantly the arbitrary dimensions. 
I definitely want to put it back in _somewhere_ , but still don't know whether it should go at the top of the file itself (like in the other examples) or into License.txt as you suggested earlier.
Maybe "inspired by" should be "derived from original code by ...".

stefanv commented on skimage/segmentation/_expand_labels.py:
```diff
@@ -0,0 +1,107 @@
+def expand_labels(label_image, distance):
+    """Expand labels in label image by ``distance`` pixels without overlapping.
+
+    Given a label image, each label is grown by up to distance pixels.
```
What does grow mean exactly?  That the border expands "outwards" by up to N pixels?  I don't think the meaning here is that you grow it only by ``distance`` pixels.

VolkerH replied:
thanks, I will try and reword this

stefanv commented on skimage/segmentation/_expand_labels.py:
```diff
@@ -0,0 +1,107 @@
+    using the option "Distance-N".
+
+    The basic idea is that you have some seed labels that you want
+    to grow by n pixels to give a mask for a larger object.
```
a repeat of the problematic description

VolkerH replied:
I reworded the whole docstring as there was some redundancy between the top part of the docstring and the `Notes` section. I think I made the description more exact and less hand-waving now, but I think this might be at the cost of being intuitively understandable. The code itself and the example result actual explain this better than I can with prose. Feel free to change the wording.

VolkerH commented:
@stefanv 
I wish I could take credit for the simplicity, but I would not have come up with this idea. The distance transform implementations in other libraries I have previously worked with did not return the location of the closest forground pixel. 
I would still have no idea `scipy.ndimage.distance_transform_edt` supports this feautre if it wasn't for peeking at the CellProfiler's Distance-N code.

Having said that, I think this is ready for a further review.

stefanv commented:
Thanks for taking a look at that docstring!  I think you're right: the new description is accurate, but less easy to follow.  I was hoping to find a description that is slightly hand-wavy but intuitive.  E.g., "grows labels outward by up to d pixels, without overflowing into neighboring labels".  I'll leave it up to you to decide how technically accurate vs. readable you want to make it.

Otherwise, looks all good to me!

VolkerH commented:
Ok. After sleeping over this I put the license string back in the `_expand_labels.py` file as this is consistent with the existing skimage functions that were derived from the Cellprofiler code base. At least some of the other examples mention _is licensed under both GPL and BSD licenses_ . However, the current `LICENSE.txt` in the CP repo only mentions BSD, so I changed that lin particular wording accordingly.

I did another rework on the docstring, attempting a more approachable description in the first sentence followed by a more exact description of the behaviour. Also when mentioning the "upstream implementation" in the Notes, I specified that this refers to `distance_transform_edt`.

I think all the outstanding comments are now addressed, so please have another look.

jni commented:
@VolkerH 

> I did another rework on the docstring, attempting a more approachable description in the first sentence followed by a more exact description of the behaviour. Also when mentioning the "upstream implementation" in the Notes, I specified that this refers to distance_transform_edt.

I think your reworkings broke the documentation build, see the CI. Having a look for culprints now...

jni commented on skimage/segmentation/_expand_labels.py:
```diff
@@ -0,0 +1,107 @@
+def expand_labels(label_image, distance):
+    """Expand labels in label image by ``distance`` pixels without overlapping.
+
+    Given a label image, `expand_labels` grows label regions (connected components) 
```
```suggestion
    Given a label image, ``expand_labels`` grows label regions (connected components) 
```

jni commented on skimage/segmentation/_expand_labels.py:
```diff
@@ -0,0 +1,107 @@
+    pixel, the label value of the closest connected component will be assigned (see 
+    Notes for the case of multiple labels at equal distance).
+     
+
```
```suggestion
```

jni commented on skimage/segmentation/_expand_labels.py:
```diff
@@ -0,0 +1,107 @@
+    Notes
+    -----
+    Where labels are spaced more than ``distance`` pixels are apart, this is
+    equivalent to a morpholgical dilation with a disc or hyperball of radius ``distance``.
```
```suggestion
    equivalent to a morphological dilation with a disc or hyperball of radius ``distance``.
```

jni commented on skimage/segmentation/_expand_labels.py:
```diff
@@ -0,0 +1,107 @@
+    However, in contrast to a morphological dilation, ``expand_labels`` will
+    not expand a label region into a neighboring region.  
+
+    This implementation of ``expand_labels`` is derived from CellProfiler [1], where
```
```suggestion
    This implementation of ``expand_labels`` is derived from CellProfiler [1]_, where
```

jni commented on skimage/segmentation/_expand_labels.py:
```diff
@@ -0,0 +1,107 @@
+    not expand a label region into a neighboring region.  
+
+    This implementation of ``expand_labels`` is derived from CellProfiler [1], where
+    it is known as module "IdentifySecondaryObjects (Distance-N)" [2].
```
```suggestion
    it is known as module "IdentifySecondaryObjects (Distance-N)" [2]_.
```

jni commented on skimage/segmentation/_expand_labels.py:
```diff
@@ -0,0 +1,107 @@
+
+    There is an important edge case when a pixel has the same distance to
+    multiple regions, as it is not defined which region expands into that
+    space. Here, the exact bahaviour depends on the upstream implementation
```
```suggestion
    space. Here, the exact bahavior depends on the upstream implementation
```

jni commented:
I've left some suggestions where there was incorrect formatting in the docstring, but I don't think off the top of my head that these should have caused errors. Anyway if I get a chance I'll try to do a local doc build.

VolkerH commented:
Thanks @jni, I need to read up on sphinx docstring syntax. 
Commited all your changes, let's see whether it makes a difference to the CI builds.

VolkerH commented:
I think I found the culprit, a single whitespace introduced in the docstring of the gallery example (indentation). Hope the build goes through now.

jni commented:
It looks happy! My original approval stands. =D

emmanuelle commented on skimage/measure/_label.py:
```diff
@@ -91,3 +90,4 @@ def label(input, neighbors=None, background=None, return_num=False,
      [0 0 0]]
     """
     return clabel(input, neighbors, background, return_num, connectivity)
+
```
```suggestion
```

emmanuelle replied:
no need for a blank line at the end of the file

emmanuelle commented on skimage/segmentation/_expand_labels.py:
```diff
@@ -0,0 +1,106 @@
+from scipy.ndimage import distance_transform_edt
+
+
+def expand_labels(label_image, distance):
```
Could we have a default value for `distance` (for example `distance=1`)? This would be more consistent with the rest of our API where most of the time the positional arguments are image arrays and all other arguments are keyword arguments.

VolkerH replied:
Thanks for the suggestion, I added a default value of 1 pixel distance.

emmanuelle commented:
Thank you very much @VolkerH ! I left a couple of small comments (the most important one is whether we want `distance` to be a kwarg with a default value of 1), and some comments of @rfezzani have not been addressed yet. But we're very close to merging I think!

VolkerH commented:
Thanks for having a look @emmanuelle, I applied your suggestions.

Thanks also for pointing out that I overlooked some of @rfezzani 's comments (Apologies! It turned out that Github had truncated those and I had to expand them to make them visible).

 All tests pass locally, so I hope the CI will also work. 

Thanks everyone for their feedback.

jni commented:
Here we go! Another excellent contribution @VolkerH! =D

asasama12 commented:
Naive question: when will this function be available in the main scikit-image python package?

grlee77 commented:
> Naive question: when will this function be available in the main scikit-image python package?

It should appear in an upcoming scikit-image 0.18 is release later this year (I don't think we have a specific date at this point).
